File: PHP/examples/examples.php
<?php /**************************************************************
**
** examples.php
**
**   Small examples that illustrate how the richreports.py module can
**   be used.
**
**   All required modules must be installed in order for this example
**   module to work correctly.
**
*/

include_once("uxadt.php"); if (!defined('_')) define('_', null);
include_once("richreports.php");

/********************************************************************
** Definition of an abstract syntax.
*/

\uxadt\_('Program', array(
    'Program' => array(array('Stmt'))
  ));

\uxadt\_('Stmt', array(
    'Repeat' => array(array('Stmt')),
    'Prints' => array('Exp'),
    'Errors' => array(),
    'Pass' => array()
  ));

\uxadt\_('Exp', array(
    'OpPlus' => array('Exp', 'Exp'),
    'OpMax' => array('Exp', 'Exp'),
    'OpAbs' => array('Exp'),
    'Num' => array('#')
  ));

/********************************************************************
** Example conversion algorithm that turns an abstract syntax tree
** into a report.
*/

function report($a) {
  if ($a->_(Program(_))) {
    list($ss) = $a();
    $rs = array(); foreach ($ss as $s) array_push($rs, report($s));
    return Page(Block(array(), array(), $rs));
  } else if ($a->_(Repeat(_))) {
    list($ss) = $a();
    $rs = array(); foreach ($ss as $s) array_push($rs, report($s));
    return
      Concat(array(
        Line(array(Keyword('repeat'))),
        Block(array(), array(), $rs)
      ));
  } else if ($a->_(Prints(_))) {
    list($e) = $a();
    return Line(array(Keyword('print'), Entity(Space()), report($e)));
  } else if ($a->_(Errors())) {
    return
      Line(array(
        Span(array(HighlightError()), array(Text('This is a message that applies to the whole highlighted span.')), array(Keyword('error'))),
        Entity(Space()), Text('outside'), Entity(Space()), Text('span')
      ));
  } else if ($a->_(Pass())) {
    return Line(array(Keyword('pass')));
  } else if ($a->_(OpPlus(_,_))) {
    list($e1,$e2) = $a();
    return Concat(array(report($e1), Keyword('+'), report($e2)));
  } else if ($a->_(OpMax(_,_))) {
    list($e1,$e2) = $a();
    return Concat(array(Builtin('max'), Punctuation('('), report($e1), Punctuation(','), report($e2), Punctuation(')')));
  } else if ($a->_(OpAbs(_))) {
    list($e) = $a();
    return Concat(array(Builtin('abs'), Punctuation('('), report($e), Punctuation(')')));
  } else if ($a->_(Num(_))) {
    list($n) = $a();
    return Atom(array(), array(Text('int')), array(Konstant(''.$n.'')));
  }
}

/********************************************************************
** Example of an abstraxt syntax tree, and rendering thereof as an
** interactive HTML report.
*/

$p = Program(array(Repeat(array(Prints(OpPlus(Num(2), Num(3))), Pass(), Errors(), Repeat(array(Prints(Num(7)), Pass())))), Prints(OpMax(OpAbs(Num(4)), Num(5))), Pass()));
echo \richreports\html(report($p));

/*eof*/ ?>

File: Haskell/Text/RichReports/Examples.hs
---------------------------------------------------------------------
-- 
-- | RichReports
--
-- @Text\/RichReports\/Examples.hs@
--
--   Examples that illustrate how the RichReports module can be
--   used.
--
--   All required modules must be installed in order for this example
--   module to work correctly.
--

---------------------------------------------------------------------
-- 

module Data.RichReports.Examples
  where

import qualified System.IO as IO (IO, putStr, writeFile)
import qualified Text.Ascetic.HTML as H

import qualified Text.RichReports as R

---------------------------------------------------------------------
-- | Definition of an abstract syntax.

data Program =
    Program [Stmt]
  deriving (Eq, Show)

data Stmt =
    Repeat [Stmt]
  | Prints Exp
  | Errors
  | Pass
  deriving (Eq, Show)
  
data Exp =
    OpPlus Exp Exp
  | OpMax Exp Exp
  | OpAbs Exp
  | Num Integer
  deriving (Eq, Show)

---------------------------------------------------------------------
-- | Example conversion algorithm that turns an abstract syntax tree
--   into a report.
  
instance R.ToReport Program where
  report (Program ss) = R.Page $ R.Block [] [] [R.report s | s <- ss]
  
instance R.ToReport Stmt where
  report s = case s of
    Repeat ss -> R.Concat [R.Line [R.Keyword "repeat", R.Block [] [] [R.report s | s <- ss]]]
    Prints e  -> R.Line [R.Keyword "print", R.Entity R.Space, R.report e]
    Errors    ->
      R.Line [
        R.Span [R.HighlightError] [R.Text "This is a message that applies to the whole highlighted span"] [R.Keyword "error"],
        R.Entity R.Space,
        R.Text "outside",
        R.Entity R.Space,
        R.Text "span"
      ]
    Pass      -> R.Line [R.Keyword "pass"]

instance R.ToReport Exp where
  report e = case e of
    OpPlus e1 e2 -> R.Concat [R.report e1, R.Keyword "+", R.report e2]
    OpMax e1 e2  -> R.Concat [R.Builtin "max", R.Punctuation "(", R.report e1, R.Punctuation ",", R.report e2, R.Punctuation ")"]
    OpAbs e      -> R.Concat [R.Builtin "abs", R.Punctuation "(", R.report e, R.Punctuation ")"]
    Num i        -> R.Atom [] [R.Text "int"] [R.Konstant (show i)]

---------------------------------------------------------------------
-- | Example of an abstraxt syntax tree, and rendering thereof as an
--   interactive HTML report.

p = Program [
      Repeat [
        Prints (OpPlus (Num 2) (Num 3)),
        Pass,
        Errors,
        Repeat [Prints (Num 7), Pass]
      ], 
      Prints (OpMax (OpAbs (Num 4)) (Num 5)),
      Pass
    ]
    
main :: IO.IO ()
main =
  do { IO.putStr $ "The program:\n" ++ (show p)
     ; IO.putStr $ "\n\nThe report:\n" ++ (show (R.report p))
     ; IO.writeFile "report.html" (show (H.html (R.report p)))
     ; IO.putStr $ "\n\nThe file \"report.html\" has been written.\n\n"
     }

--eof